### PR TITLE
Rmb 385 add 'query index name' to schema.json and allow compound indexes

### PR DESCRIPTION
**Pull request by evaluk**

### CODE REVIEW

File: cql2pgjson/src/main/java/org/folio/cql2pgjson/util/Cql2SqlUtil.java [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 import java.util.regex.Pattern;
 import org.folio.cql2pgjson.exception.QueryValidationException;
+import org.folio.rest.persist.ddlgen.Index;
```
Comment (julianladisch) [inline cql2pgjson/src/main/java/org/folio/cql2pgjson/util/Cql2SqlUtil.java]: Can you remove this import?

File: dbschema/src/main/java/org/folio/rest/persist/ddlgen/Table.java [hunk 2]
```diff
@@ -1,5 +1,6 @@
 package org.folio.rest.persist.ddlgen;
 
+import java.util.ArrayList;
```
Comment (julianladisch) [inline dbschema/src/main/java/org/folio/rest/persist/ddlgen/Table.java]: Can you remove this import?

File: domain-models-runtime/src/main/resources/templates/db_scripts/general_functions.ftl [hunk 3]
```diff
@@ -110,3 +110,8 @@ BEGIN
   RETURN NEW;
 END;
 $$ language 'plpgsql';
+
+
+create or replace function concat_space_sql(VARIADIC text[])
+RETURNS text AS $$ select concat_ws(' ', VARIADIC $1); 
+$$ LANGUAGE SQL IMMUTABLE;
```
Comment (julianladisch) [inline domain-models-runtime/src/main/resources/templates/db_scripts/general_functions.ftl]: Can you explain on this pull request why concat_space_sql has been choosen in favor of text_concat_ws?

File: domain-models-runtime/src/test/resources/templates/db_scripts/compoundIndexTest.json [hunk 4]
```diff
@@ -0,0 +1,46 @@
+{
```
Comment (julianladisch) [inline domain-models-runtime/src/test/resources/templates/db_scripts/compoundIndexTest.json]: Can you rename compoundIndexTest.json to compoundIndex.json?

File: domain-models-runtime/src/test/java/org/folio/rest/persist/ddlgen/SchemaMakerTest.java [hunk 5]
```diff
@@ -47,6 +47,29 @@ public void canCreateAuditedTable() throws IOException, TemplateException {
...
+  public void canCreateSQLExpressionIndex() throws IOException, TemplateException {
+    SchemaMaker schemaMaker = new SchemaMaker("harvard", "circ", TenantOperation.UPDATE,
+      "mod-foo-18.2.3", "mod-foo-18.2.4");
+    String json = ResourceUtil.asString("templates/db_scripts/compoundIndexTest.json");
+    schemaMaker.setSchema(ObjectMapperTool.getMapper().readValue(json, Schema.class));
+    String result = schemaMaker.generateDDL();
+
+    assertThat(result, containsString("lower(concat_space_sql(jsonb->>'field1', jsonb->>'field2'))"));
+
+  }
```
Comment (julianladisch) [inline domain-models-runtime/src/test/java/org/folio/rest/persist/ddlgen/SchemaMakerTest.java]: Can you add an empty line after the method?

File: domain-models-runtime/src/test/java/org/folio/rest/persist/ddlgen/SchemaMakerTest.java [hunk 6]
```diff
@@ -47,6 +47,29 @@ public void canCreateAuditedTable() throws IOException, TemplateException {
...
+  @Test
+  public void canCreateSQLExpressionIndex() throws IOException, TemplateException {
+    SchemaMaker schemaMaker = new SchemaMaker("harvard", "circ", TenantOperation.UPDATE,
+      "mod-foo-18.2.3", "mod-foo-18.2.4");
+    String json = ResourceUtil.asString("templates/db_scripts/compoundIndexTest.json");
+    schemaMaker.setSchema(ObjectMapperTool.getMapper().readValue(json, Schema.class));
+    String result = schemaMaker.generateDDL();
+
+    assertThat(result, containsString("lower(concat_space_sql(jsonb->>'field1', jsonb->>'field2'))"));
+
```
Comment (julianladisch) [inline domain-models-runtime/src/test/java/org/folio/rest/persist/ddlgen/SchemaMakerTest.java]: Can you remove this empty line?

File: domain-models-runtime/src/test/java/org/folio/rest/persist/ddlgen/SchemaMakerTest.java [hunk 7]
```diff
@@ -47,6 +47,29 @@ public void canCreateAuditedTable() throws IOException, TemplateException {
...
+  public void canCreateCompoundIndex() throws IOException, TemplateException {
+    SchemaMaker schemaMaker = new SchemaMaker("harvard", "circ", TenantOperation.UPDATE,
+      "mod-foo-18.2.3", "mod-foo-18.2.4");
+    String json = ResourceUtil.asString("templates/db_scripts/compoundIndexTest.json");
+    schemaMaker.setSchema(ObjectMapperTool.getMapper().readValue(json, Schema.class));
+    String result = schemaMaker.generateDDL();
+
+    assertThat(result, containsString("concat_space_sql(tablea.jsonb->>'firstName' , tablea.jsonb->>'lastName')"));
+
+  }
```
Comment (julianladisch) [inline domain-models-runtime/src/test/java/org/folio/rest/persist/ddlgen/SchemaMakerTest.java]: Can you remove this empty line?

File: domain-models-runtime/src/test/resources/templates/db_scripts/compoundIndexTest.json [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+          "multiFieldNames": "firstName,lastName",
+          "tOps": "ADD",
+          "caseSensitive": false,
+          "removeAccents": false
+        }
+      ],
+      "fullTextIndex": [
+        {
+          "fieldName": "ftfield",
+           "multiFieldNames": "field1,field2",
```
Comment (julianladisch) [inline domain-models-runtime/src/test/resources/templates/db_scripts/compoundIndexTest.json]: Can you fix the indentation?

File: domain-models-runtime/src/test/java/org/folio/rest/persist/ddlgen/SchemaMakerTest.java [hunk 9]
```diff
@@ -47,6 +47,29 @@ public void canCreateAuditedTable() throws IOException, TemplateException {
...
 
+  @Test
+  public void canCreateCompoundIndex() throws IOException, TemplateException {
+    SchemaMaker schemaMaker = new SchemaMaker("harvard", "circ", TenantOperation.UPDATE,
+      "mod-foo-18.2.3", "mod-foo-18.2.4");
+    String json = ResourceUtil.asString("templates/db_scripts/compoundIndexTest.json");
+    schemaMaker.setSchema(ObjectMapperTool.getMapper().readValue(json, Schema.class));
+    String result = schemaMaker.generateDDL();
+
+    assertThat(result, containsString("concat_space_sql(tablea.jsonb->>'firstName' , tablea.jsonb->>'lastName')"));
```
Comment (julianladisch) [inline domain-models-runtime/src/test/java/org/folio/rest/persist/ddlgen/SchemaMakerTest.java]: Can you also add a check for the fullTextIndex?

Comment (julianladisch) [inline domain-models-runtime/src/test/java/org/folio/rest/persist/ddlgen/SchemaMakerTest.java]: Can you also check that schemaMaker.getSchema().getTables().get(0).get*Index().get(0).createCompoundIndex("tablea") is correct?

File: domain-models-runtime/src/test/java/org/folio/rest/persist/ddlgen/SchemaMakerTest.java [hunk 10]
```diff
@@ -47,6 +47,29 @@ public void canCreateAuditedTable() throws IOException, TemplateException {
...
+
+  @Test
+  public void canCreateSQLExpressionIndex() throws IOException, TemplateException {
+    SchemaMaker schemaMaker = new SchemaMaker("harvard", "circ", TenantOperation.UPDATE,
+      "mod-foo-18.2.3", "mod-foo-18.2.4");
+    String json = ResourceUtil.asString("templates/db_scripts/compoundIndexTest.json");
+    schemaMaker.setSchema(ObjectMapperTool.getMapper().readValue(json, Schema.class));
+    String result = schemaMaker.generateDDL();
+
+    assertThat(result, containsString("lower(concat_space_sql(jsonb->>'field1', jsonb->>'field2'))"));
```
Comment (julianladisch) [inline domain-models-runtime/src/test/java/org/folio/rest/persist/ddlgen/SchemaMakerTest.java]: Can you also check for address?

Comment (julianladisch) [inline domain-models-runtime/src/test/java/org/folio/rest/persist/ddlgen/SchemaMakerTest.java]: Can you also check that schemaMaker.getSchema().getTables().get(*).get*Index().get(*).createCompoundIndex("tablea") is correct?

File: cql2pgjson/src/test/resources/compoundIndex.sql [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+END; $$ language 'plpgsql';
+CREATE TRIGGER update_tableb_references BEFORE INSERT OR UPDATE ON tableb
+  FOR EACH ROW EXECUTE PROCEDURE update_tableb_references();    
+INSERT INTO tablea (jsonb) VALUES
+('{"id": "A0000000-0000-0000-0000-000000000000", "firstName": "Mike", "lastName": "Smith","field1": "first0", "field2": "last0"}'),
+('{"id": "A1111111-1111-1111-1111-111111111111", "firstName": "Tom", "lastName": "Jones","field1": "first1", "field2": "last1"}'),
+('{"id": "A2222222-2222-2222-2222-222222222222", "firstName": "Lucy", "lastName": "Williams","field1": "first2", "field2": "last2"}'),
+('{"id": "A3333333-3333-3333-3333-333333333333", "firstName": "Anne", "lastName": "Davis","field1": "first3", "field2": "last3"}'),
+('{"id": "A4444444-4444-4444-4444-444444444444", "firstName": "Mary", "lastName": "Miller","field1": "first4", "field2": "last4"}');
+INSERT INTO tablea (jsonb) VALUES (jsonb_build_object('id', md5('a' || generate_series(1, 2000)::text)));
```
Comment (julianladisch) [inline cql2pgjson/src/test/resources/compoundIndex.sql]: Can you remove this additional generated than cannot influence the test?

File: cql2pgjson/src/test/resources/compoundIndex.sql [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+INSERT INTO tableb (jsonb) VALUES
+('{"id": "B1111111-1111-1111-1111-111111111111", "city": "Boston", "state": "MA","field1": "first0", "field2": "last0" }'),
+('{"id": "B2222222-2222-2222-2222-222222222222", "city": "Chicago", "state": "IL","field1": "first1", "field2": "last1"}'),
+('{"id": "B3333333-3333-3333-3333-333333333333", "city": "San Francisco", "state": "CA","field1": "first2", "field2": "last2"}'),
+('{"id": "B4444444-4444-4444-4444-444444444444",  "city": "Austin", "state": "TX","field1": "first3", "field2": "last3"}');
+INSERT INTO tableb (jsonb) VALUES (jsonb_build_object('id', md5('b' || generate_series(1, 2000)::text)));
```
Comment (julianladisch) [inline cql2pgjson/src/test/resources/compoundIndex.sql]: Can you remove this additional generated than cannot influence the test?

File: dbschema/src/main/java/org/folio/rest/persist/ddlgen/Index.java [hunk 13]
```diff
@@ -55,5 +57,36 @@ public String getArraySubfield() {
...
+  public void setMultiFieldNames(String queryIndexName) {
+    this.multiFieldNames = queryIndexName;
+  }
+  public String getSqlExpression() {
+    return sqlExpression;
+  }
+  public void setSqlExpression(String sqlExpression) {
+    this.sqlExpression = sqlExpression;
+  }
+  public String createCompoundIndex(String tableLoc) {
```
Comment (julianladisch) [inline dbschema/src/main/java/org/folio/rest/persist/ddlgen/Index.java]: Can you rename it to getFinalSqlExpression(String) table? fieldPath is not compound.

**Comment by julianladisch** [comment]
Can you add a unit test with a nested field, for example proxy.personal.firstName and proxy.personal.lastName?

File: domain-models-runtime/src/test/java/org/folio/rest/persist/ddlgen/SchemaMakerTest.java (hunk 9 above)
Comment (evaluk) [reply domain-models-runtime/src/test/java/org/folio/rest/persist/ddlgen/SchemaMakerTest.java]: The naming convention here is due to the choice of using the sql native function with a fixed  concat.  A new function is needed to guarantee the immutable value of the function for use in the index.

File: domain-models-runtime/src/main/resources/templates/db_scripts/general_functions.ftl (hunk 3 above)
Comment (evaluk) [reply domain-models-runtime/src/main/resources/templates/db_scripts/general_functions.ftl]: The choice of name here is based on the need for an immutable function to use in indexes.  The function is native sql and defines the concat character as space.

**Pull request merged**